### DISCUSSION TITLE
<Link/> vs <a href> in Slate

### DISCUSSION THREAD

**Discussion by zehawki** [react-page/react-page#1130] [Q&A]
Currently Slate creates an `<a href>` for all links - internal or external. This of course leads to a nasty headache of the site reloading on click even when its an internal link (which should have been smoothly handled by the router). So the solution is to instantiate a `<Link/>` for internal and `<a href>` for external.

So I added to my slate customization the following:

```
    // https://react-page.github.io/docs/#/slate?id=example-2-customize-the-inbuilt-slate-plugins
    link: {
      anchor: def.plugins.link.anchor,
      // link: def.plugins.link.link,
      link: def.plugins.link.link(linkDef => ({
        ...linkDef, // spread it, so that the new config contains all defaults
        // Component: ({ children, openInNewWindow, href, attributes }) => {
        Component: (props) => {

          // Logic for <Link/> creation based on internal, external, anchor etc etc

        }
      })),
    },
```
But the problem is Slate editing of the link. With the above, it stays a link even in editing mode and a click on it will lose all the page edits. I looked deeply into how Slate is working for link editing and came up with this to mirror what it does without my customization:

```
    // https://react-page.github.io/docs/#/slate?id=example-2-customize-the-inbuilt-slate-plugins
    link: {
      anchor: def.plugins.link.anchor,
      // link: def.plugins.link.link,
      link: def.plugins.link.link(linkDef => ({
        ...linkDef, // spread it, so that the new config contains all defaults
        // Component: ({ children, openInNewWindow, href, attributes }) => {
        Component: (props) => {

          // console.log(props);

          // This is when its being edited! The link should not work, hence this method is adopted
          // This has come after lot of experimentation to mirror how Slate seems to be handling this
          if (props.attributes["data-slate-node"] === "element")
            return (
              <a data-slate-node="element" data-slate-inline="true" href={props.href}>
                {props.children}
              </a>
            )

          else {
            // Logic for <Link/> creation based on internal, external, anchor etc etc

          }

        }
      })),
    },
```

@macrozone could you please confirm if there is a more elegant way to do this, I think this method is clunky, or?

PS: Shouldn't this functionality be part of the library?

**Comment by macrozone** [comment]
> So the solution is to instantiate a `<Link/>` for internal and `<a href>` for external.

absolut correct!

> But the problem is Slate editing of the link. With the above, it stays a link even in editing mode and a click on it will lose all the page edits.

yes, had the same problem, I solved it outside of react-page, but was thinking about providing a helper for that:

```
// create a context
const InEditorContext = React.createContext(false);

// wrap the editor with it:
// if you want to render it in read-only, set the value to false or do not wrap it with the context provider at all

<InEditorContext.Provider value={true}>
     <Editor  /* .... */ />
 </InEditorContext.Provider>


// now you can create a hook to check whether you are in the editor

// returns true when in editor
const useIsInEditor = () => {
  return useContext(InEditorContext);
};

```

so in your link component, you could prevent the click when `useIsInEditor` returns true. 

I added this helper hook for that:

```
const usePreventOnClickInEditor = () => {
  const isInEditor = useIsInEditor();

  return isInEditor ? (e) => e.preventDefault() : undefined;
};
```

and then i assigned this to `<Link />` components:

```
const onClick = usePreventOnClickInEditor() 

// later

<Link onClick={onClick} href={href} />

```


so what we could add is a hook that returns true when in readOnly mode.
